torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -994,57 +997,70 @@ struct to_ir {
   }
 
   void emitReturn(const Return& stmt) {
-    TypePtr result_type = def_stack_.back().declared_return_type_;
-    Value* result = emitExpr(stmt.expr(), result_type);
+    Value* actual_return = emitExpr(stmt.expr());
+    TypePtr actual_return_type = actual_return->type();
+    TypePtr declared_return_type = def_stack_.back().declared_return_type_;
     // result type is annotated, every return must convert to that type
-    if (result_type) {
+    if (declared_return_type) {
       // this guard skips implicit conversion from None -> Tensor for the return
       // type. otherwise forgetting a return a function returning a tensor will
       // cause a None to be converted to a tensor.
-      if (!(result_type->isSubtypeOf(TensorType::get()) &&
-            result->type()->isSubtypeOf(NoneType::get()))) {
-        result = tryConvertToType(
+      if (!(declared_return_type->isSubtypeOf(TensorType::get()) &&
+            actual_return_type->isSubtypeOf(NoneType::get()))) {
+        actual_return = tryConvertToType(
             stmt.range(),
             *graph,
-            result_type,
-            result,
+            declared_return_type,
+            actual_return,
             /*allow_conversions=*/true);
       }
-
-      if (!result->type()->isSubtypeOf(result_type)) {
-        throw ErrorReport(stmt.range())
-            << "Return value was annotated as having type "
-            << result_type->repr_str() << " but is actually of type "
-            << result->type()->repr_str();
+      if (!actual_return_type->isSubtypeOf(declared_return_type)) {
+        // Throw if 1) we don't have a Union, or 2) the Union does not
+        // hold the actual return type.
+        if (declared_return_type->kind() != UnionType::Kind ||
```

Comment:
I am probably missing something. Isn't this condition necessary for line 1017 to be true? I think as long as line 1017 is true, we should always throw regardless of whether `union` is involved or not.